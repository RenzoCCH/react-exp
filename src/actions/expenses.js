import database from "../firebase/firebase";

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns object
// component dispatches function (?)
// redux store changes (has the ability to dispatch other actions and do whatever it wants)

//ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense: expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };
    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

//EDIT_EXPENSE
export const editExpense = expense => ({
  type: "EDIT_EXPENSE",
  expense
});

export const startEditExpense = expense => {
  const expenseData = {
    amount: expense.amount,
    createdAt: expense.createdAt,
    description: expense.description,
    note: expense.note
  };
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${expense.id}`)
      .update(expenseData)
      .then(() => {
        dispatch(editExpense(expense));
      });
  };
};

//SET_EXPENSES
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenses = [];
    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then(snapshot => {
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
