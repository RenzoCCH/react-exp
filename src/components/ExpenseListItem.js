import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

export const ExpenseListItem = ({ id, description, amount, createdAt }) => {
  return (
    <Link className="list-item" to={`/edit/${id}`}>
      <div>
        <h3 className="list-item__title">{description}</h3>
        {createdAt !== 0 && (
          <span className="list-item__subtitle">{moment(createdAt).format("MMMM Do, YYYY")}</span>
        )}
      </div>
      <h3 className="list-item__data">
        {amount !== 0 && <span>{numeral(amount / 100).format("$0,0.00")}</span>}
      </h3>
    </Link>
  );
};

export default ExpenseListItem;
