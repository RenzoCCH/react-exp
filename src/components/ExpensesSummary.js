import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from "../selectors/expenses";
import totalExpenses from "../selectors/expenses-total";
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <p>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</p>
  );
};


const mapStateToProps = (state) => {
  let expenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: expenses.length,
    expensesTotal: totalExpenses(expenses)
  };
};



export default connect(mapStateToProps)(ExpensesSummary);
