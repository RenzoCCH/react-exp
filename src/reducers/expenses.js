//Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState,action = {})=>{
  switch (action.type) {
  case 'ADD_EXPENSE':
    return [...state,action.expense];
  case 'REMOVE_EXPENSE':
    return state.filter(({id})=>action.id!==id);
  case 'EDIT_EXPENSE':
    return state.map((exp)=>{
      if(exp.id === action.expense.id){
        return {...exp, ...action.expense};
      }
      else return exp;
    });
  default:
    return state;
  }
};

