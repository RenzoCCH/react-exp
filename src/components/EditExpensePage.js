import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onEditExpense = expense => {
    this.props.startEditExpense({ ...expense, id: this.props.expense.id });
    this.props.history.push("/");
  };

  onRemoveExpense = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onEditExpense}
          />
          <button className="button--secondary" onClick={this.onRemoveExpense}>Remove Expense</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: expense => dispatch(startEditExpense(expense)),
    startRemoveExpense: data => dispatch(startRemoveExpense(data))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
