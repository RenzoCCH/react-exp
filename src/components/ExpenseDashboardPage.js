import React from 'react';
import ExpenseList from './ExpenseList.js';
import ExpenseListFilters from './ExpenseListFilters.js';
import ExpensesSummary  from './ExpensesSummary.js';


const ExpenseDashboardPage = ()=>(
  <div>
    <ExpenseListFilters/>
    <ExpensesSummary/>
    <ExpenseList/>
  </div>
);

export default ExpenseDashboardPage;