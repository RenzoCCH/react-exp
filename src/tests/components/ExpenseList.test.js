import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render eExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render expenseList with empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render expenseListItem', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} key={expenses[0].id}/>);
  expect(wrapper).toMatchSnapshot();
});

//const ExpenseList = (props) =>{
//  return (
//  <div>
//    <h1>Expense List</h1>
//    {props.expenses.map((expense)=>(
//      <ExpenseListItem
//        {...expense}
//        key={expense.id}/>))}
//  </div>
//)};
//
//const mapStateToProps = (state)=>{
//  return {
//    expenses: selectExpenses(state.expenses,state.filters)
//  };
//}
//
//export default connect(mapStateToProps)(ExpenseList);
//
