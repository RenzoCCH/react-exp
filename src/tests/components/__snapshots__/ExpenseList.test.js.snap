// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render eExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={195}
      createdAt={"1970-01-01T00:00:00.000Z"}
      description="Gum"
      id="1"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={109500}
      createdAt={"1969-12-28T00:00:00.000Z"}
      description="Rent"
      id="2"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={4500}
      createdAt={"1970-01-05T00:00:00.000Z"}
      description="Credit Card"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;

exports[`should render expenseList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`should render expenseListItem 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Gum
    </h3>
    <span
      className="list-item__subtitle"
    >
      December 31st, 1969
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    <span>
      $1.95
    </span>
  </h3>
</Link>
`;
